cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
-            for k in key:
-                if not isinstance(k, NumPyThunk):
-                    raise NotImplementedError(
-                        "need support for mixed advanced indexing"
+            index_map = []
+
+            # First, we need to check if transpose is needed
+            for dim, k in enumerate(key):
+                if np.isscalar(k) or isinstance(k, NumPyThunk):
+                    if start_index == -1:
+                        start_index = dim
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
+            else:
+                index_map = tuple(range(len(key)))
+
+            for d, k in enumerate(key):
+                dim = index_map[d]
+                if np.isscalar(k):
+                    if k < 0:
+                        k += store.shape[dim + shift]
+                    store = store.project(dim + shift, k)
+                    shift -= 1
+                elif k is np.newaxis:
+                    store = store.promote(dim + shift, 1)
+                elif isinstance(k, slice):
+                    store = store.slice(dim + shift, k)
+                elif isinstance(k, NumPyThunk):
+                    if k.dtype == np.bool:
+                        if k.shape[0] != store.shape[dim]:
+                            raise ValueError(
+                                "boolean index did not match "
+                                "indexed array along dimension  "
+                            )
+                        # in case of the mixed indises we all nonzero
+                        # for the bool array
+                        k = k.nonzero()
+                        tuple_of_arrays += k
+                    else:
+                        tuple_of_arrays += (self.runtime.to_deferred_array(k),)
+                else:
+                    raise TypeError(
+                        "Unsupported entry type passed to advanced",
+                        "indexing operation",
                     )
-                tuple_of_arrays += (k,)
+            if store.transformed:
+                # after store is transformed we need to to return a copy of
+                # the store since Copy operation can't be done on
+                # the store with transformation
+                rhs, store = self._copy_store(store)
         else:
```

Comment:
In the case of a single integer index array, can't you just fall back to the tuple case? I think you could structure the code like this:

```
if isinstance(key, NumPyThunk) and key.dtype != np.bool:
  key = (key,)
if isinstance(key, tuple):
  ...
else:
  assert isinstance(key, NumPyThunk) and key.dtype != np.bool
  ...
```

I think if you structure it like that, you can keep just the boolean case from this branch.

Reply:
Actually, maybe the cleanest way to structure the code is like so:

```
if isinstance(key, NumPyThunk) and key.dtype == np.bool and key.shape == rhs.shape:
  # special boolean case
  return False, rhs, out
if isinstance(key, NumPyThunk):
  key = (key,)
assert isinstance(key, tuple)
...
```

Then you don't even have to do anything special for the case where `key.dtype == np.bool` but `key.shape != rhs.shape`; it will already be handled by the tuple case.

Reply:
Will do, thank you